src/intptrcast.rs

```diff
@@ -72,12 +74,45 @@ impl<'mir, 'tcx> GlobalStateInner {
                     None
                 }
             }
-        };
-        // Pointers created from integers are untagged.
-        Pointer::new(
-            alloc_id.map(|alloc_id| Tag { alloc_id, sb: SbTag::Untagged }),
-            Size::from_bytes(addr),
-        )
+        }?;
+
+        match global_state.provenance_mode {
+            ProvenanceMode::Legacy => Some(alloc_id),
+            ProvenanceMode::Permissive if global_state.exposed.contains(&alloc_id) =>
+                Some(alloc_id),
+            _ => None,
+        }
+    }
+
+    pub fn expose_addr(ecx: &MiriEvalContext<'mir, 'tcx>, alloc_id: AllocId) {
+        trace!("Exposing allocation id {:?}", alloc_id);
+
+        let mut global_state = ecx.machine.intptrcast.borrow_mut();
+        global_state.exposed.insert(alloc_id);
```

Comment:
Let's skip this unless the provenance mode is "Permissive"; it's just a waste of time otherwise.